pkg/controller/autoscaler/metrics_client.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package autoscalercontroller
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/expapi"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+
+	heapster "k8s.io/heapster/api/v1/types"
+)
+
+// An interface for getting metrics for pods.
+type MetricsClient interface {
+	ResourceConsumption(namespace string) ResourceConsumptionClient
+}
+
+type ResourceConsumptionClient interface {
+	// Gets average resource consumption for pods under the given selector.
+	Get(resourceName api.ResourceName, selector map[string]string) (*expapi.ResourceConsumption, error)
+}
+
+// Aggregates results into ResourceConsumption. Also returns number of
+// pods included in the aggregation.
+type metricAggregator func(heapster.MetricResultList) (expapi.ResourceConsumption, int)
+
+type metricDefinition struct {
+	name       string
+	aggregator metricAggregator
+}
+
+// Heapster-based implementation of MetricsClient
+type HeapsterMetricsClient struct {
+	client client.Interface
+}
+
+type HeapsterResourceConsumptionClient struct {
+	namespace           string
+	client              client.Interface
```

Comment:
Hrm - would probably be better to only include the things we need, so
having both interfaces (PodNamespacer, ServiceNamespacer) might be better
in the short run.  But it's more of a desire to be specific about what we
need, so I'm ok not making this change.

On Mon, Aug 31, 2015 at 1:29 PM, Marcin Wielgus <notifications@github.com>
wrote:

> In pkg/controller/autoscaler/metrics_client.go
> <https://github.com/kubernetes/kubernetes/pull/13303#discussion_r38329906>
> :
>
> > +// pods included in the aggregation.
> > +type metricAggregator func(heapster.MetricResultList) (expapi.ResourceConsumption, int)
> > +
> > +type metricDefinition struct {
> > +	name       string
> > +	aggregator metricAggregator
> > +}
> > +
> > +// Heapster-based implementation of MetricsClient
> > +type HeapsterMetricsClient struct {
> > +	client client.Interface
> > +}
> > +
> > +type HeapsterResourceConsumptionClient struct {
> > +	namespace           string
> > +	client              client.Interface
>
> Not only, I also call heapster through the proxy interface.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13303/files#r38329906>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift